hir/src/asm.rs

```diff
@@ -0,0 +1,2447 @@
+    /// and loads the word at that address to the top of the stack.
+    pub fn loadw(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemLoadw);
+    }
+
+    /// Loads the word at the given address to the top of the stack.
+    pub fn loadw_imm(self, addr: u32) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemLoadwImm(addr));
+    }
+
+    /// Pops two elements, the first containing a memory address from the top of the stack,
+    /// the second the value to be stored as the first element of the word at that address.
+    pub fn store(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemStore);
+    }
+
+    /// Pops an element from the top of the stack, and stores it as the first element of
+    /// the word at the given address.
+    pub fn store_imm(self, addr: u32) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemStoreImm(addr));
+    }
+
+    /// Pops two elements, the first containing a memory address + element offset from the
+    /// top of the stack, the second the value to be stored to the word at that address,
+    /// using the offset to determine which element will be written to.
+    pub fn store_offset(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemStoreOffset);
+    }
+
+    /// Pops an element from the top of the stack, and stores it at the given offset of
+    /// the word at the given address.
+    pub fn store_offset_imm(self, addr: u32, offset: u8) {
+        assert!(
+            offset < 4,
+            "invalid element offset, must be in the range 0..=3, got {}",
+            offset
+        );
+        self.stack.drop();
+        self.asm
+            .push(self.ip, MasmOp::MemStoreOffsetImm(addr, offset));
+    }
+
+    /// Pops an element containing a memory address from the top of the stack,
+    /// and then pops a word from the stack and stores it as the word at that address.
+    pub fn storew(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemStorew);
+    }
+
+    /// Pops a word from the stack and stores it as the word at the given address.
+    pub fn storew_imm(self, addr: u32) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemStorewImm(addr));
+    }
+
+    /// Pops a boolean value from the stack, and executes the first block if it is true,
+    /// otherwise the second block.
+    pub fn if_true(self, then_blk: MasmBlockId, else_blk: MasmBlockId) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::If(then_blk, else_blk))
+    }
+
+    /// Pops a boolean value from the stack, and executes the given block if it is true,
+    /// otherwise it is skipped. The given block will continue to execute for as long as
+    /// the top value on the stack at the end of the block is true.
+    pub fn while_true(self, body: MasmBlockId) {
+        self.stack.drop();
```

Comment:
In general we can't assume that the stack height will be the same at this program point in every iteration of the loop, so simply dropping the conditional value doesn't look correct.

My guess is that you intend to make it a requirement that each program point has a fixed stack height, but that requirement is not specified anywhere, and it puts a limitation on the inline assembly that the programmer can write, so it should be documented somewhere.

Reply:
I'm taking the liberty of unresolving this conversation to make a point:

Arguments along the lines of "it makes little sense for anyone to write such code" scare me in the context of smart contracts, because the code we're compiling may end up handling billions of dollars worth of assets, and any undefined behavior can potentially be used by malicious actors.

In other words, the space for what code makes sense to write is very broad indeed, and is certainly not limited to code that has well-defined behavior.

Reply:
> Arguments along the lines of "it makes little sense for anyone to write such code" scare me in the context of smart contracts, because the code we're compiling may end up handling billions of dollars worth of assets, and any undefined behavior can potentially be used by malicious actors.

Let me be more precise. We aren't responsible for someone writing MASM by hand that throws a billion dollars into the equivalent of a crypto incinerator. That is entirely the responsibility of the person writing the code. Inline assembly is by definition an escape hatch, you are being given virtually unlimited power to shoot yourself in the foot however you like. Undefined behavior isn't saying "do this at your own risk", it's a declaration that you _should never do that thing_. Indeed Miden already has a number of instructions for which undefined behavior is specified - it doesn't mean we don't know what happens when, e.g., an unchecked u32 operation overflows past 2^32; it means we don't know how that will affect the rest of the program.

In any case, the issue raised here is moot - the purpose of the loop/if builders that have been added is to validate the inline assembly upon construction, and at least catch if the operand stack was improperly maintained in a way that breaks assumptions that the rest of the compiler makes about the behavior of a given inline assembly block (i.e. the number and types of elements are mismatched across different exit paths in the control flow graph). But we can't make any guarantees about the behavior of the inline assembly itself - it might _look_ like things are in sync, but it may very well be that the author has the same logical value in two different positions on the operand stack along one path vs another. We can't do anything about that.

However, if _we_ introduce undefined behavior, i.e. we lower something with well-defined semantics, such as checked math, to instructions which do not preserve those semantics, and may lead to undefined behavior - that's a different thing altogether. We are responsible for preserving the semantics and behavior that we document, no more, no less. If we document that use of a specific instruction requires certain guarantees on its use, and that failing to uphold those guarantees is undefined behavior, that isn't an issue or a bug, that is expected for a compiler IR.